##  <u>Lab 4 : Group 20</u>


**Overall Objective**: Create a model that predicts rent prices as well as possible for typical New York City apartments.

### Part 0


 - **<u>Group Number:</u>** 20
 - **<u>Group Members:</u>**<br>
 Pallavi Ravikumar Menon<br>
 Akshara Santhosh

     

### Part 1 - Create and evaluate an initial model



#### 1. A peek into the data

In [1]:
##importing libraries
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error, r2_score

In [2]:
#reading the data
df = pd.read_csv("C:/Users/User/Desktop/ML/rent.csv")
df.head().T

0  \
bathrooms                                                      1.0   
bedrooms                                                         2   
building_id                                                      0   
created                                           21-06-2016 07:29   
description      Makr this stunning place your new home!- 2 bed...   
display_address                                           New Lots   
features                            ['Pre-War', 'Hardwood Floors']   
latitude                                                   40.6678   
longitude                                                 -73.8788   
manager_id                        17dd8baf35198562ff9524b486f1754a   
photos           ['https://photos.renthop.com/2/7194943_5fafcb8...   
price                                                         1475   
street_address                                        864 New Lots   
interest_level                                                high   
num_desc_words                                                  41   

                                                                 1  \
bathrooms                                                      1.0   
bedrooms                                                         2   
building_id                       3cc586423f9e977a0f991b010242baa6   
created                                           22-06-2016 06:44   
description      Amazing location!Hardwood floors marble bath g...   
display_address                                    Bleecker Street   
features         ['Elevator', 'Laundry in Building', 'Dishwashe...   
latitude                                                   40.7285   
longitude                                                    -74.0   
manager_id                        1fb46c4a72bcf764ac35fc23f394760d   
photos           ['https://photos.renthop.com/2/7201657_9fe4dcc...   
price                                                         3800   
street_address                                 160 Bleecker Street   
interest_level                                                high   
num_desc_words                                                  46   

                                                                 2  \
bathrooms                                                      1.0   
bedrooms                                                         1   
building_id                       cd759a988b8f23924b5a2058d5ab2b49   
created                                           04-05-2016 02:26   
description      DON'T MISS OUT !!! ONE IN A KIND GORGEOUS, HUG...   
display_address                                           E 56 St.   
features         ['Doorman', 'Elevator', 'Laundry in Building',...   
latitude                                                   40.7575   
longitude                                                 -73.9626   
manager_id                        f52612ad89c61a21e84172f0af0d4436   
photos           ['https://photos.renthop.com/2/6962322_dc4fbd1...   
price                                                         2850   
street_address                                        405 E 56 St.   
interest_level                                                 low   
num_desc_words                                                 119   

                                                                 3  \
bathrooms                                                      1.0   
bedrooms                                                         3   
building_id                       250ae11aef7ca5668ea55763085d3ff8   
created                                           29-06-2016 04:31   
description      Not the prettiest, but large, bright and cheap...   
display_address                                        E 69 Street   
features                                                        []   
latitude                                                   40.7657   
longitude                                                 -73.9572   
ma

In [3]:
df["interest_level"] = df["interest_level"].map({"low":1,"medium":2,"high":3})
df.head().T

0  \
bathrooms                                                      1.0   
bedrooms                                                         2   
building_id                                                      0   
created                                           21-06-2016 07:29   
description      Makr this stunning place your new home!- 2 bed...   
display_address                                           New Lots   
features                            ['Pre-War', 'Hardwood Floors']   
latitude                                                   40.6678   
longitude                                                 -73.8788   
manager_id                        17dd8baf35198562ff9524b486f1754a   
photos           ['https://photos.renthop.com/2/7194943_5fafcb8...   
price                                                         1475   
street_address                                        864 New Lots   
interest_level                                                   3   
num_desc_words                                                  41   

                                                                 1  \
bathrooms                                                      1.0   
bedrooms                                                         2   
building_id                       3cc586423f9e977a0f991b010242baa6   
created                                           22-06-2016 06:44   
description      Amazing location!Hardwood floors marble bath g...   
display_address                                    Bleecker Street   
features         ['Elevator', 'Laundry in Building', 'Dishwashe...   
latitude                                                   40.7285   
longitude                                                    -74.0   
manager_id                        1fb46c4a72bcf764ac35fc23f394760d   
photos           ['https://photos.renthop.com/2/7201657_9fe4dcc...   
price                                                         3800   
street_address                                 160 Bleecker Street   
interest_level                                                   3   
num_desc_words                                                  46   

                                                                 2  \
bathrooms                                                      1.0   
bedrooms                                                         1   
building_id                       cd759a988b8f23924b5a2058d5ab2b49   
created                                           04-05-2016 02:26   
description      DON'T MISS OUT !!! ONE IN A KIND GORGEOUS, HUG...   
display_address                                           E 56 St.   
features         ['Doorman', 'Elevator', 'Laundry in Building',...   
latitude                                                   40.7575   
longitude                                                 -73.9626   
manager_id                        f52612ad89c61a21e84172f0af0d4436   
photos           ['https://photos.renthop.com/2/6962322_dc4fbd1...   
price                                                         2850   
street_address                                        405 E 56 St.   
interest_level                                                   1   
num_desc_words                                                 119   

                                                                 3  \
bathrooms                                                      1.0   
bedrooms                                                         3   
building_id                       250ae11aef7ca5668ea55763085d3ff8   
created                                           29-06-2016 04:31   
description      Not the prettiest, but large, bright and cheap...   
display_address                                        E 69 Street   
features                                                        []   
latitude                                                   40.7657   
longitude                                                 -73.9572   
ma

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 15 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   bathrooms        20000 non-null  float64
 1   bedrooms         20000 non-null  int64  
 2   building_id      20000 non-null  object 
 3   created          20000 non-null  object 
 4   description      19367 non-null  object 
 5   display_address  19951 non-null  object 
 6   features         20000 non-null  object 
 7   latitude         20000 non-null  float64
 8   longitude        20000 non-null  float64
 9   manager_id       20000 non-null  object 
 10  photos           20000 non-null  object 
 11  price            20000 non-null  int64  
 12  street_address   19997 non-null  object 
 13  interest_level   20000 non-null  int64  
 14  num_desc_words   20000 non-null  int64  
dtypes: float64(3), int64(4), object(8)
memory usage: 2.3+ MB


#### 2. creating the model

In [5]:
#Selecting only numeric coloumns from the dataset.
df_num = df[['bathrooms', 'bedrooms', 'longitude','latitude','interest_level','num_desc_words', 'price']]
df_num.head(2)

bathrooms  bedrooms  longitude  latitude  interest_level  num_desc_words  \
0        1.0         2   -73.8788   40.6678               3              41   
1        1.0         2   -74.0000   40.7285               3              46   

   price  
0   1475  
1   3800

In [6]:
#checking the null values
print(df_num.isnull().sum()) 
#There are no null values

bathrooms         0
bedrooms          0
longitude         0
latitude          0
interest_level    0
num_desc_words    0
price             0
dtype: int64


In [7]:
X= df_num.drop('price', axis=1)
y= df_num['price']


In [8]:
#splitting up the data,where 20% of data goes into test set, 80% into training set
X_train, X_val, y_train, y_val = train_test_split(X,y, test_size=0.2) 

In [9]:
rf = RandomForestRegressor(n_estimators=100, n_jobs = 1) 
rf.fit(X_train,y_train)
r2_train= rf.score(X_train, y_train).round(4)
r2_val = rf.score(X_val,y_val).round(4)

print( f"R2 value for training:{r2_train:.4f}")
print( f"R2 value for validation:{r2_val:.4f}")


R2 value for training:0.9331
R2 value for validation:0.0165


In [10]:
#Evaluation of the model
rf = RandomForestRegressor(n_estimators=100, n_jobs = 1,oob_score=True) 
rf.fit(X_train,y_train)

noisy_oob_r2 = rf.oob_score_
print(f"OOB score is {noisy_oob_r2:.4f}")

OOB score is 0.5687


#### 3. Verifying Mean Absolute Error

In [11]:
train_r2 = []
train_mae = []
val_r2 =[]
val_mae=[]
oob_scores =[]

for i in range(10):
    X_train, X_val, y_train, y_val = train_test_split(X,y, test_size=0.2)
    rf.fit(X_train,y_train)
    train_pred = rf.predict(X_train)
    val_pred = rf.predict(X_val)
    train_r2.append(round(r2_score(y_train, train_pred),2))
    train_mae.append(round(mean_absolute_error(y_train, train_pred),2))
    val_r2.append(round(r2_score(y_val, val_pred),2))
    val_mae.append(round(mean_absolute_error(y_val, val_pred),2))
    oob_scores.append(rf.oob_score_)

In [12]:
print("The scores of Training R-square are: \n", train_r2)
print(" ")
print("Validation R-square scores are: \n", val_r2)
print(" ")
print("The Training MAE scores are: \n", train_mae)
print(" ")
print("The Validation MAE scores are: \n", val_mae)
print(" ")
print("The Out-of-bag scores are:  \n", oob_scores)

The scores of Training R-square are: 
 [0.85, 0.8, 0.94, 0.72, 0.94, 0.94, 0.8, 0.76, 0.83, 0.76]
 
Validation R-square scores are: 
 [-10.02, 0.3, 0.01, 0.51, 0.01, 0.01, 0.38, -2.64, -3.92, -20.3]
 
The Training MAE scores are: 
 [254.28, 246.26, 193.49, 250.58, 194.19, 193.14, 253.44, 257.41, 252.82, 258.13]
 
The Validation MAE scores are: 
 [674.56, 560.52, 811.04, 590.41, 760.95, 780.64, 560.89, 616.27, 572.19, 777.14]
 
The Out-of-bag scores are:  
 [-0.7056608703753291, -0.37199154247639443, 0.5398361725018448, -0.2816550296107927, 0.5772575188996807, 0.5660923191661924, -0.5160002030598945, -0.16534539924443803, -0.28678413429926564, -0.5801425563010374]


In [13]:
#Finding out the mean of r2, MAE of both training and validation data and oob_score
print("Mean train r2: ", np.mean(train_r2).round(3)) 
print("Mean validation r2: ", np.mean(val_r2).round(3))
print("Mean train MAE: ", np.mean(train_mae).round(3))
print("Mean validation MAE: ", np.mean(val_mae).round(3))
print("Mean oob score: ", np.mean(oob_scores).round(3))

Mean train r2:  0.834
Mean validation r2:  -3.566
Mean train MAE:  235.374
Mean validation MAE:  670.461
Mean oob score:  -0.122


### <u> Explanation for part 1</u>

1. Importing the necessary libraries used in this Lab.

2. Understanding the datasets :<br>
    -Got familiar with the datatypes of each column.<br>
    -Checked for the null values.

3. Transformed categorical ordinal column "Interest_level" to numeric ordinal value to include in the model training.

4. Selected numeric columns for the final dataset and proceeded with splitting the dataset into Train and Validation set.

5. Fit the data in the RandomForestRegressor and evaluated the R2 value for both the set.

6. There is a significant difference between the R2 values of both the Training and Validation set.

7. The oob score was calculated for the same set which came around to be 0.5687.

8. Trained the datsaset 10 times to fit into the regressor model and calculated the MAE values for both the sets.

9. The mean values for the calculations is found out having larger difference for R2 values.

10. The oob score was found out to be negative which is very worse for our model. It shows that there might be a possibility of noise in the data.





### Part 2 - Denoise the data


In [28]:
df_num.describe()

bathrooms      bedrooms     longitude      latitude  interest_level  \
count  20000.000000  20000.000000  20000.000000  20000.000000    20000.000000   
mean       1.214050      1.545250    -73.826650     40.703469        1.386750   
std        0.500695      1.114188      6.681871      1.802293        0.628963   
min        0.000000      0.000000   -124.015100    -18.979000        1.000000   
25%        1.000000      1.000000    -73.991600     40.728300        1.000000   
50%        1.000000      1.000000    -73.977900     40.751700        1.000000   
75%        1.000000      2.000000    -73.954700     40.774500        2.000000   
max        6.000000      7.000000    158.395000     49.343000        3.000000   

       num_desc_words         price  
count    20000.000000  2.000000e+04  
mean      -962.539300  3.733582e+03  
std      32390.385788  8.014783e+03  
min   -1000000.000000 -1.800000e+04  
25%         48.000000  2.500000e+03  
50%         81.000000  3.150000e+03  
75%        118.000000  4.134000e+03  
max        518.000000  1.070000e+06

#### Cleaning the data for denoising after studying the description

In [15]:
#cleaning up the data of bedrooms and bathrooms
df_clean =df_num[(df_num['bedrooms'] <= 5) | (df_num['bathrooms'] <= 1)&(df_num['bathrooms']>=4.5)]

In [16]:
 #cleaning up the location
df_clean =df_clean[(df_clean['longitude']!= 0) | (df_clean['latitude']!= 0)]
df_clean = df_clean[(df_clean['latitude']>40.55) &
                        (df_clean['latitude']<40.84) &
                        (df_clean['longitude']>-74.1) &
                        (df_clean['longitude']<-73.67)]

In [17]:
 #cleaning up the price
df_clean = df_clean[(df_clean['price'] > 1000) & (df_clean['price'] < 11000)]

In [18]:
#cleaning on Num_desc_words
df_clean =df_clean[(df_clean['num_desc_words'] <= 15)]

In [21]:
#checking the impact on number of samples after denoising

print(f"The impact on number of samples:{(df_num.shape[0] - df_clean.shape[0]) / df_num.shape[0]}")
print(f"Number of cancelled rows:{(df_num.shape[0] - df_clean.shape[0])}")


The impact on number of samples:0.9033
Number of cancelled rows:18066


**After performing the denoising on the dataset, 18066 out of 20000 were removed as to filter the dataset for the final model thereby decreasing the inconsistencies present in the dataset.**

### Part 3 - Create and evaluate a final model


In [22]:
x_clean = df_clean.drop('price', axis=1)
y_clean = df_clean['price']

In [23]:
x_clean_train, x_clean_val, y_clean_train, y_clean_val = train_test_split(x_clean, y_clean, test_size=0.2)  
rf_clean = RandomForestRegressor(n_estimators=100, n_jobs=-1, oob_score=True) 
rf_clean.fit(x_clean_train, y_clean_train) 
r2_clean_train= rf_clean.score(x_clean_train, y_clean_train).round(4)
r2_clean_val = rf_clean.score(x_clean_val, y_clean_val).round(4)
print( f"R2 value for clean training data  :{r2_clean_train:.4f}")
print( f"R2 value for clean validation data:{r2_clean_val:.4f}")

R2 value for clean training data  :0.9580
R2 value for clean validation data:0.7281


In [24]:
rf_clean.oob_score_.round(4)

0.7323

In [25]:
train_r2 = []
train_mae = []
val_r2 = []
val_mae = []
oob_scores = []

for i in range(10):
    x_clean_train, x_clean_val, y_clean_train, y_clean_val = train_test_split(x_clean, y_clean, test_size=0.2)
    rf_clean = RandomForestRegressor(n_estimators=100, n_jobs=-1, oob_score=True) 
    rf_clean.fit(x_clean_train, y_clean_train)
    train_preds = rf_clean.predict(x_clean_train)
    val_preds = rf_clean.predict(x_clean_val)
    train_r2.append(round(r2_score(y_clean_train, train_preds), 4))
    val_r2.append(round(r2_score(y_clean_val, val_preds), 4))
    train_mae.append(round(mean_absolute_error(y_clean_train, train_preds), 4))
    val_mae.append(round(mean_absolute_error(y_clean_val, val_preds), 2))
    oob_scores.append(round(rf_clean.oob_score_,4))


In [26]:
print("The Scores after removing the Noise are: \n")
print("The Training R-square scores: \n", train_r2)
print("")
print("The Validation R-Square scores: \n", val_r2)
print("")
print("The Training  MAE scores: \n", train_mae)
print("")
print("The Validation MAE scores: \n", val_mae)
print("")
print("The Out-of-bag scores are:  \n", oob_scores )

The Scores after removing the Noise are: 

The Training R-square scores: 
 [0.9586, 0.9581, 0.9576, 0.959, 0.958, 0.9585, 0.961, 0.9581, 0.9619, 0.958]

The Validation R-Square scores: 
 [0.7238, 0.7572, 0.7598, 0.6871, 0.7562, 0.7314, 0.7314, 0.7787, 0.6793, 0.7601]

The Training  MAE scores: 
 [197.4915, 198.3428, 199.0141, 202.3786, 201.4943, 195.9913, 193.2157, 195.6356, 192.3729, 195.4508]

The Validation MAE scores: 
 [477.02, 475.87, 480.13, 467.53, 463.67, 486.53, 490.46, 478.08, 530.61, 486.03]

The Out-of-bag scores are:  
 [0.7231, 0.7192, 0.7199, 0.7234, 0.7257, 0.7338, 0.7308, 0.7241, 0.7384, 0.733]


In [27]:
# mean on r2, MAE,oob_score after denoising the data
print("Mean train r2: ", round(np.mean(train_r2), 4))
print("Mean validation r2: ", round(np.mean(val_r2), 4))
print("Mean train MAE: ", round(np.mean(train_mae), 2))
print("Mean validation MAE: ", round(np.mean(val_mae), 2))
print("Mean oob score: ", round(np.mean(oob_scores),4))

Mean train r2:  0.9589
Mean validation r2:  0.7365
Mean train MAE:  197.14
Mean validation MAE:  483.59
Mean oob score:  0.7271


### <u> Explanation for part 3</u>

1. Further analysis and removing noise from the data a clean dataset was created .
2. The model was trained after splitting the data in the ratio 80:20.
3. Same algorithm of code was followed to fit in and calculate the R2 values of both the sets.
4. We can see the value of R2 obtained in part 1 increases to about 0.7281.
5. The MAE values for both training and validation set were calculated and the difference obtained part 1 was reduced by almost approximately 250.
6. The oob score observed for the final model was 0.7271 compared to negative oob score of -0.122 .
7. Overall, the significant increase in oob score value makes it apparent that model is close to better after denoising the dataset.



### Part 4 - Document the problems 

In this part, please use the table below to document your understanding of all the data issues you discovered. Note that **no code** should be included, as that should be covered in **Part 2**. Also, note that even if one line of code fixed a few problems, you should list each problem separately in the table below, so be sure you have investigated the data properly. For example, if the list `[-6, 5, 0, 50]` represents heights of adults, the -6, 0, and 50 would represent three data issues to be included in the table below, even though one line of code may be able to address all of them. 

| Data issue discovered | Why is this a problem? | How did you fix it? | Why is this fix appropriate? |
| :- | :- | :- | :- | 
|  Some data points having bedroom > 5 |   For a typical New York apartment very selective residents would have bedroom greater than 5  |We specified range using comparison and logical operators   | It removes the possible outliers i.e residents having greater number of bedrooms thereby generalizing the dataset more.  |
|  Bathrooms less than 0 and more than 4.5 | Possibility of apartments having no bathroom is minimal and apartments having more than 4.5 bathroom is considered luxury    | We specified range using comparison and logical operators |The dataset further gets refined and filters apartment that are considered typical for new york city.   |
|Latitude and Longitude were found to be 0 |  0 latitude and 0 longitude specifics location that are at the equators. |Excluded data entries having latitude and longitude as 0  | Location is further defined as these values are removed.  |
|  Irrelevant Latitude and Longitude  |   Coordinates not representing New York city is not useful in this context |Specified the range of latitude and longitute values representing new york city cooordinates   | The dataset further gets filtered accurately as per the desired location.  |
|  Price Variation  |   There were extremities in price ranges : negative values in some and values upto 1 million |After observing the dataset using desc() we were able to specify price ranges.| Variation in prices had massive impact over the mean value of prices.  |
|Number of words in the description |  Certain description having 0 to 1 words possibly were apartments having less information and not legitimate  | Range of word count was decided for a typical apartment description   | It removes the possibility of unwanted apartment listing without proper information.  |